### ISSUE TITLE
Breaking change incorrectly reported for a stable package with a multi-user concurrent workflow

### ISSUE THREAD

**Issue by dylan-bourque** [bufbuild/buf#3552]
### GitHub repository with your minimal reproducible example (do not leave this field blank or fill out this field with "github.com/bufbuild/buf" or we will automatically close your issue, see the instructions above!)

https://github.com/dylan-bourque/buf-breaking-repro

### What's up?

The `buf breaking` command can incorrectly report a breaking change when different users separately make independent, non-breaking changes to a stable proto package.  The linked repo demonstrates the problems.

The scenario is:
Given a stable proto package, 2 developers each create a new branch (`feature1` and `feature2`).  Each makes a non-breaking change in their branch.  In the linked example the change is adding a new message.

After one new feature is merged, the developer working on the other will start to see an incorrect failure from `buf breaking` reporting a deleted message when comparing against the `main` branch in the Git remote.

```
> buf breaking --against "https://github.com/dylan-bourque/buf-breaking-repro.git#branch=main,subdir=proto"
proto/foo/bar/v1/foobar.proto:1:1:Previously present message "Baz" was deleted from file.
```

The `Baz` message was actually **added** by the other feature but a direct comparison between the `main` and the working branch will treat that diff as a delete.  While the issue can easily be corrected by doing a forward merge (or rebase) to update the open feature branch with the new changes from `main`, it would be nice if this didn't require manual intervention.

Some potential solutions:
1. synthesize a temporary workspace that contains the working branch with any changes from `main` merged in and compare _that_ to `main`
2. find the commit that the working branch was created from and translate the reference in `--against` to use _that_ instead of `main`
3. update the comparison algorithm to do a 3-way comparison between the working branch, `main` and the common ancestor to eliminate changes made outside of the working branch

Option 1 was suggested by someone in the Buf Slack when I asked about this when I first noticed it a few months ago.  Option 3 would be the most flexible, but I suspect also the most complicated to implement.

**Comment by emcfarlane** [comment]
Hi @dylan-bourque if this is about GitHub actions in CI you can use the `pull_request` [GitHub event ](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request) event to avoid this issue. Using this event type results in the merged state of the target branch. So this would avoid the error of this missing field.

If this is not about the CI: I don't think it's safe to alter the input `.` to be a merged state based on `git` as that would need to be applied to all inputs, which it can't be.

**Comment by nicksnyder** [comment]
To clarify, the context of this issue is about using the Buf CLI in CI, but without assuming any particular code host like GitHub, and also not assuming usage of the BSR.

For reasons that Ed mentioned (e.g. interactions with the BSR), I don't think this is a bug in the CLI, what it is doing is a reasonable/straightforward (and conservative) interpretation of what is being requested. That said, I think we want to support your use case too.

The current CLI already allows you to explicitly provide a ref to compare against, so you could do something like this to achieve the behavior of solution #2:
```
buf breaking --against "https://github.com/dylan-bourque/buf-breaking-repro.git#branch=main,subdir=proto,ref=$(git merge-base main HEAD)"
```

If we were to support a different behavior in the CLI, I think we would need to have a new option or flag to avoid breaking compatibility.

**Comment by dylan-bourque** [comment]
@nicksnyder I think using `ref=...` is a reasonable workaround but IMO it would be a better UX to provide an option to do that directly within the CLI command.

Either way, I think it would be a nice touch to at least document this scenario and call out either the option or the workaround.

**Comment by emcfarlane** [comment]
Using `ref=git merge-base main HEAD` would give the parent commit and potentially cause false negatives. For example on the [buf-breaking-repro](https://github.com/dylan-bourque/buf-breaking-repro) we have:
```
     feature1
    /
---1---main
```
Where the merge base of `main feature1` is `1`. This would miss any breaking changes merged from `1..main`

I think we want to compare against the merged state of `feature1` into `main`:
```
     feature1
    /        \
---1---main---M
```
We would then compare `M` to `main`. For the buf-breaking-repo example this would fail as a merge conflict currently exists. For other cases where the branches are mergeable this would avoid the deleted field issues seen and ensure no breaking changes. To do this behaviour I think we would need to specify the input to be merged against main. This could be done with an new option `merge=<ref>` which would merge the ref specified (effectively `git merge --no-commit --no-ff <ref>`) and the do the comparison against. The command would look like:
```
buf breaking .git#merge=main --against=".git#branch=main"
```

**Comment by nicksnyder** [comment]
Agree that comparing against a merged state is ideal.

It isn't clear to me what we would want/need to specify a ref value for the merge parameter. Wouldn't we always want the ref to match the ref provided in the --against flag?

```
buf breaking .git#merge--against=".git#branch=main"
```

**Comment by dylan-bourque** [comment]
I agree that comparing against a merged baseline would be the most correct but it also feels like separating the "and create a merged state to compare against" option is clunky.  As @nicksnyder points out, it creates scenarios where a user would have to specify the branch/ref twice.  Add in the possibility of typos and the UX would be arguably worse than just using the `ref=...` workaround.

Could the new option be added to the syntax of the `--against` like `buf breaking --against=".git#branch=main,merge" (or something along those lines)?

**Comment by emcfarlane** [comment]
Currently inputs don't depend on other flags or arguments. The command `buf breaking --against=.git#branch=main` says take the implicit input `.`, of type [dir](https://buf.build/docs/reference/inputs/#dir), and compare it against `.git#branch=main`, of type [git](https://buf.build/docs/reference/inputs/#git). The directory input isn't aware the `--against` is a git input, and the git input doesn't know it's being compared a local directory. Adding a `merge` option without a `<ref>` would require thinking through how the `ref` is resolved:
- Is the `against` input required to be a git input? Presumably yes, so we could error when `--against` is of another type.
- What about a remote, or even different repository? `buf breaking .git#merge --against="https://github.com/unrelated/repo`
- Does the merge option fail for other commands without an `--against`?

Specifying the `merge=<ref>` allows the input option to be specified for all commands, e.g. `buf build .git#merge=main` would check the module can build when merged into main. The input would be useable outside of `buf breaking`. Agree that the UX of specifying the `<ref>` twice is a sore spot, and the benefit of `merge` option outside breaking is limited.

**Comment by emcfarlane** [comment]
Documentation has been added to [buf.build/docs/ci-cd/setup](https://buf.build/docs/ci-cd/setup/#handling-concurrent-non-breaking-changes) to show how to implement the merge behaviour in git for both CI and local script setups. Discussing the tradeoffs of adding the option to the Buf CLI we believe this workaround should suit most use-cases. Closing to resolve.
